core/src/replay_stage.rs

```diff
@@ -2368,6 +2368,13 @@ impl ReplayStage {
                 last_voted_slot
             );
         }
+
+        // If we are a non voting validator or have an incorrect setup preventing us from
+        // generating vote txs, no need to refresh
+        if tower.last_vote_tx_blockhash() == Hash::default() {
```

Comment:
This one is definitely kind of tricky. Namely, for nodes that are not submitting valid votes, it seems like most of the fields in `Tower` will be valid as if the node were actually landing votes. But not this field.

Wonder if there is a way we can convey that with types. I see `last_vote_tx_blockhash` is decorated with `#[serde(skip)]` in `Tower` so we can change without worrying about compatibility; would it make sense to change this field to an option ? Or if we don't want to touch the fields of `Tower`, move this check into the accessor
```rust
    pub fn last_vote_tx_blockhash(&self) -> Option<Hash> {
        Some(self.last_vote_tx_blockhash).filter(|hash| hash != Hash::default())
    }
```
It doesn't look like there are currently any other callers of `Tower::last_vote_tx_blockhash`, but by moving the logic into `Tower`, we'll avoid burning ourselves if someone else begins calling this function

Reply:
oo didn't see the `[serde(skip)]`. Should be safe to rework this as an `Option`, hate that we use `Hash::default()` in so many places already.